---
title: "PhyloProfile: a R package for phylogenetic profiling"
author: "\\
Yulong Niu \\

College of Life Sciences, Sichuan University, China"
date: "`r Sys.Date()`"
bibliography: ppRef.bib
csl: nature.csl
output:
  BiocStyle::html_document:
    toc: true
  BiocStyle::pdf_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{PhyloProfile: a R package for phylogenetic profiling}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r style, echo=FALSE, results='asis', message=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(tidy = FALSE,
                      warning = FALSE,
                      message = FALSE)
```

``` {r macro, echo=FALSE, results='hide', message=FALSE}
phyloprofile <- function() {"*[PhyloProfile](https://github.com/YulongNiu/PhyloProfile)*"}
Robject <- function(x){sub('%obj%', x, '<span style="background-color:#F0F0F0;color:#404040;font-family:\'Lucida Console\', monospace">%obj%</span>')}
Rclass <- function(x){sub('%obj%', x, '<span style="font-family:\'Times New Roman\', Times, serif;font-style: italic">%obj%</span>')}
```

# Abstract

`r phyloprofile()` contains various analysis methods and visualization of phylogenetic profiles. 

## Preprocess

+ Retrieving phylogenetic profile from the [STRING](http://string-db.org/) database
+ Normalization 
    * Normalized Phylogenetic Profile (NPP)
    * Singular Value Decomposition (SVD)
   
## Supported analysis

+ Co-occurrency methods
    * Jaccard similarity
    * Pearson correlation coefficient
    * mutual information
    * Hamming distance
    * Euclidean distance
    * co-occurrency method with collapsed tree
+ Model-based methods
    * Maximum likelihood
    * Dollo parsimony distance
+ Downstream analysis
    * Expanding known gene-sets
    * Significant gene-set analysis
    
## Visualization

+ Phylogenetic profile plot
+ correlation matrix plot
+ [Circos](http://circos.ca/) plot
+ [D3](https://d3js.org/) network

# Introduction

Phylogenetic profiling is a bioinformatic technique used to show the presence and absence of homologous across a large number of species. This genomic-context method is based on the assumption that functionally related proteins are likely to been evolved in a correlated pattern [@pellegrini1999assigning]. It has been successfully applied to detect novel physical protein-protein interactions, new components of biological complexes, and potential linkages in signaling pathways or metabolic processes [@kensche2008practical].

As increasing numbers of sequenced genomes have become available, various phylogenetic profiling online tools and databases have been developed [@cheng2015protphylo; @sadreyev2015phylogene; @ott2012napp; @szklarczyk2014string; @cromar2016phylopro2]. However, systematically analysis of user-defined phylogenetic profile on local computers is still a slow and tedious process. Here, we provide a R package named `r phyloprofile()` for analyzing and visualizing phylogenetic profiles.

# Construct phylogenetic profiles

In `r phyloprofile()`, a profile with $n$ proteins across $m$ organisms is denoted as a $n \times m$ matrix containing either binning or continuous data. A binning profile uses 0 and 1 to represent the homologous absence and presence, respectively. For example, if the value in the $i$-th row and $j$-th column is 1, it tells the $i$-th protein has corresponding homologous in the $j$-th organism. A continuous profile usually consists of specific sequence similarities, for example BLAST bit scores, which need to be normalized before further analysis. 

## Read profiles from local files

Normally, a profile can be easily stored as a numeric matrix. In `r phyloprofile()` a new class named `r Rclass('PP')` is created to manipulate a profile, and `PP()` is designed to coerce a regular matrix to a `r Rclass('PP')` object. The unique row names of a profile representing protein IDs are particularly important, because the following analysis use them as indices. If row names are missing, `PP()` will automatically creates them, *e.g.* `protein1` and `protein2`, however meaningful names such as [ensembl ID](http://useast.ensembl.org/index.html) or [NCBI gene ID](https://www.ncbi.nlm.nih.gov/gene) are highly recommended. Here, we read in a published profile [@barker2007constrained] called `r Robject('secP')`, which is originally stored in a csv-formed file. In this vignette, we used some "pipe-like" operators, *e.g.* `%>%` and `%<>%` from the `r CRANpkg('magrittr')` package to improve readability of 
codes.

```{r}
library('magrittr')
library('ape')
library('PhyloProfile')

pPath <- system.file('extdata', 'bioinfoProfile.csv', package = 'PhyloProfile')
sceMat <- pPath %>% read.csv(row.names = 1) %>% as.matrix
sceP <- PP(sceMat)

## binning profile
sceP
```

The `r Rclass('PP')` class is directly inherited from the `r Rclass('matrix')` class, so the extractor `[`, `rownames()`/`colnames()`, and `rbind()`/`cbind()` work well for a `r Rclass('PP')` object. Moreover, `PPData()` is used to extract and replace the profile data.


``` {r eval=FALSE}
## profile of top ten proteins
sceP[1:10, ]

## query rownames
rownames(sceP)

## extract and replace profile
testP <- sceP
PPData(testP)
PPData(testP) <- sample(0:1, 5 * 20) %>% matrix(nrow = 5)
```

## Retrieve profiles from the STRING database

The [STRING](http://string-db.org/) database (version 10) provides raw BLAST bit scores of the best hit for more than 2,000 organisms, and it is perfectly suitable for generating raw bit score profiles with the `r Biocpkg('STRINGdb')` package. The `get_STRING_species()` gives a glance of supported organisms in STRING.

```{r}
library('STRINGdb')
sp <- get_STRING_species(version = '10', species_name = NULL)
head(sp)
```

The bit score profile that is a continuous profile can be obtained by `PPSTRING()` according to the taxonomy ID (`r Robject('species_id')` in the above outputs), for example `r Robject('9606')` for *Homo sapiens*. In the following example, we retrieve the profile of 5 human proteins. Set `idx = 'all'` for a complete human profile, and increase `n` to accelerate the download process if multiple cores are supported.

``` {r eval=FALSE}
## A continuous profile
PPSTRING(speID = 9606, idx = 1:5, n = 2)
```

## Normalize raw bit score profiles

Two normalization methods, SVD[@franceschini2015svd; @psomopoulos2013detection] and NPP[@tabach2013identification; @tabach2013human; @sadreyev2015phylogene], are provided for the pre-process of raw bit score profiles.

The algorithm of NPP for a $n \times m$ profile $P$:

1. For each bit score, it is set as `bitReset` if it lower than the `bitCutoff`.

2. For each protein (row), if its number of homologous (with bit score > `bitReset`) across $m$ organisms is lower than `mimConserve`, the protein is removed because of its poor conservation.

3. Normalizing proteins: for the $i$-th row, the bit score $p_{ij}$ is normalized as $\log_2(p_{ij}/p_{maxi})$, where $p_{maxi}$ is the maximum bit score in the $i$-th row.

4. Normalizing species (columns): for the $j$-th column, the bit score $p_{ij}$ is normalized as $(p_{ij} - \mu_j)/\sigma_j$, which is also known as the [z-score](https://en.wikipedia.org/wiki/Standard_score), where $\mu_j$ and $\sigma_j$ is the mean and the standard deviation of the $j$-th column, respectively. Compare to the original profile $P$, the NPP normalized profile $P'$ ($n' \times m$) has the same organism number $m$, but may contain less proteins.

The algorithm of SVD for a $n \times m$ profile $P$:

1. The first step is the same as the one in NPP.

2. Normalizing proteins: for the $i$-th row, the bit score $p_{ij}$ is normalized as $p_{ij}/p_{maxi}$, where $p_{maxi}$ is the maximum bit score in the $i$-th row.

3. [SVD](https://en.wikipedia.org/wiki/Singular_value_decomposition) of the profile that $\boldsymbol{P = U{\Sigma}V'}$, where $\boldsymbol{U}$ is the unitary matrix. The profile $P'$ then is defined as the top `trimming` (30% for example) columns of $\boldsymbol{U}$. 

4. Similar to the second step of NPP, poor conserved proteins are marked in the original profile $P$, and removed in $P'$ ($n' \times m'$).

5. [Euclidean normalization](https://en.wikipedia.org/wiki/Norm_(mathematics)) of species in $P'$: for the $i$-th row, the bit score $p_{ij}$ is normalized as $p_{ij}/\sqrt{\sum_{i=1}^{m'}p_{ij}^2}$. The SVD normalized profile $P'$ may have less organisms and proteins than those in the original profile $P$.


``` {r eval=FALSE}
## simulation of a bit score profile with 2000 proteins and 1000 organisms
ppRawBit <- sample(0:200, 2000 * 1000, replace = TRUE) %>% matrix(nrow = 2000) %>% PP

## NPP
Norm(ppRawBit, method = 'NPP', bitCutoff = 60, bitReset = 1, minConserve = 12)

## SVD
Norm(ppRawBit, method = 'SVD', bitCutoff = 100, bitReset = 0, minConserve = 12, trimming = 0.8)
```

# Linkage analysis

## Prepare linkages

The basic phylogenetic profiling is to analyze the relationships of paired proteins, which is also known as "linkages". The `r Robject('linkM')` below is a 2-column matrix, of which each row contains a pair of protein IDs. `PPIdx()` is used to prepare and store linkage indices corresponding to the profile `r Robject('sceP')` into a `r Rclass('PPIdx')` class object `r Robject('sceL')`. For example, the linkage `r Robject('"6323698" --> "10383785"')` is converted to indices `r Robject('4232 --> 203')`. Note that linkages containing protein IDs not in the profile are automatically removed.

``` {r}
## randomly generate 10 linkages
linkM <- sceP %>% rownames %T>% {set.seed(12345)} %>% sample(2 * 10) %>% matrix(ncol = 2)
sceL <- PPIdx(sceP, linkM)
sceL
```

For a batch of interested proteins, `PPIdx()` also facilities to generate linkages within these proteins or linkages with the rest whole proteome. In this case, `PPIdx()` accepts protein indices like `r Robject('pIdx')` below, **but not protein IDs**. Please consider set `bigmat = TRUE` if the linkage number is large.


``` {r eval=FALSE}
## indices of interested proteins, for example 5 random proteins from `sceP`
p <- c("6321976", "6321639", "6323007", "6320376", "10383799")
pIdx <- p %>% match(rownames(sceP))

## within linkages
PPIdx(sceP, pIdx, pIdx)

## with self linkages, for example "6321976" --> "6321976"
PPIdx(sceP, pIdx, pIdx, self = TRUE)

## considering linkage directions
## for example "6321976" --> "6323007" and "6323007" --> "6321976" are different
PPIdx(sceP, pIdx, pIdx, bidirect = TRUE)

## linkages with rest whole proteome 
PPIdx(sceP, pIdx, 1:nrow(sceP), bigmat = TRUE)

## whole proteome linkages
PPIdx(sceP, 1:nrow(sceP), 1:nrow(sceP), bigmat = TRUE)
```

## Combine profiles with phylogenetic trees

The phylogenetic profiling can be enhanced by combining other types of biological data, and the phylogenetic tree is widely used to trace the evolutionary processes (reviewed by Kensche *et.al.* [@kensche2008practical]). In `r phyloprofile()`, `PPTreeIdx()` is used to combine the a profile, a [rooted tree](https://www.ncbi.nlm.nih.gov/Class/NAWBIS/Modules/Phylogenetics/phylo9.html), and linkage indices into a new class `r Rclass('PPTreeIdx')` object. `r Robject('sceTree')` is the phylogenetic tree, which is a `r Rclass('phylo')` class object from the `r CRANpkg('ape')` package. Note that when combined with a tree, the organism of the profile missing in the tree are removed.

``` {r}
sceTree <- system.file('extdata', 'bioinfoTree.nex', package = "PhyloProfile") %>% read.nexus
sceTreeL <- PPTreeIdx(sceL, sceTree)
sceTreeL
```

## Co-occurrency methods

Five common-used co-occurrency methods [@pellegrini1999assigning; @kensche2008practical; @glazko2004detection] are supported to calculated similarity or distance of linkages by `SimDist()`. For each pair of proteins in a $n \times m$ profile $P$, for example $X$ and $Y$:

* `method = 'SimJaccard'` for Jaccard similarity

$$
J(X, Y) = \frac{\left\lvert\{i{\mid}x_i=1{\cap}y_i=1\}\right\rvert}
{\left\lvert\{i{\mid}x_i=1{\cup}y_i=1\}\right\rvert}
$$

* `method = 'SimCor'` for Pearson correlation coefficient

$$
cor(X, Y) = \frac{\sum\limits_{i=1}^m(x_i-\overline{X})(y_i-\overline{Y})}
{(m-1)S_XS_Y}
$$

where $\overline{X}$ and $\overline{Y}$ are the sample means of $X$ and $Y$, and $S_X$ and $S_Y$ are the sample standard deviations of $X$ and $Y$, respectively.

* `method = 'SimMI'` for mutual information

$$
I(X, Y) = \int_Y\int_Xp(x, y)\log\left(\frac{p(x, y)}{p(x)p(y)}\right)\mathrm{d}x\mathrm{d}y
$$

where $p(x, y)$ is the joint probability density function of $X$ and $Y$, $p(x)$ and $p(y)$ are the marginal probability density functions of $X$ and $Y$, respectively.

* `method = 'DistHamming'` for Hamming distance (Manhattan distance for continuous profiles)

$$
H(X, Y) = \sum\limits_{i=1}^m\lvert x_i-y_i \rvert
$$

* `method = 'DistEulidean'` for Euclidean distance 

$$
Eu(X, Y) = \sqrt{\sum\limits_{i=1}^m(x_i - y_i)^2}
$$

All the methods work well for a binning profile, whereas `SimCor` and `DistEulidean` are recommended for the NPP and SVD normalized profiles, respectively. `head()`/`tail()`/`as.data.frame()` are used to extract profiling results.

``` {r}
## mutual information with two cores
## increase n to accelerate profiling 
sceLMI <- SimDist(sceL, method = 'SimMI', n = 2)
head(sceLMI)
```

## 

## Model-based methods


## Downstream analysis

# Case studies



# Misc.

## Convert STRING protein ID

Protein ID (row names) of profiles retrieved from the STRING database are stored in a special format `r Robject('specieID.EnsemblID')`. The following scripts use the `r Biocpkg('biomaRt')` package to convert STRING IDs to other formats such as Entrez ID.

```{r eval=FALSE}
library('PhyloProfile')
library('magrittr')
library('biomaRt')

## STRING profile
hsa5p <- PPSTRING(speID = 9606, idx = 1:5, n = 2)

## step 1: extract Esembl ID
esemblIDs <- hsa5p %>%
  rownames %>%
  strsplit(split = '.', fixed = TRUE) %>%
  sapply('[', 2)

## step 2: check supported formats
ensembl <- useMart('ensembl', dataset = 'hsapiens_gene_ensembl')
listFilters(ensembl)

## step 3: convert to entrez ID
Anno <- getBM(attributes = c('ensembl_peptide_id', 'entrezgene'),
              filters = 'ensembl_peptide_id',
              value = esemblIDs,
              mart = ensembl)
```

## Objects and methods in *PhyloProfile*


# Homepage

Please visit [PhyloProfile homepage](https://github.com/YulongNiu/PhyloProfile) for more information.


# Session Information

Here is the output of `sessionInfo()` on the system on which this document was compiled:

```{r echo=FALSE}
sessionInfo()
```

# References
