---
title: "PhyloProfile: a R package for phylogenetic profiling"
author: "\\
Yulong Niu \\

College of Life Sciences, Sichuan University, China"
date: "`r Sys.Date()`"
bibliography: ppRef.bib
csl: nature.csl
output:
  BiocStyle::html_document:
    toc: true
  BiocStyle::pdf_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{PhyloProfile: a R package for phylogenetic profiling}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r style, echo=FALSE, results='asis', message=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(tidy = FALSE,
                      warning = FALSE,
                      message = FALSE)
```

``` {r macro, echo=FALSE, results='hide', message=FALSE}
phyloprofile <- function() {"*[PhyloProfile](https://github.com/YulongNiu/PhyloProfile)*"}
Robject <- function(x){sub('%obj%', x, '<span style="background-color:#F0F0F0;color:#404040;font-family:\'Lucida Console\', monospace">%obj%</span>')}
Rclass <- function(x){sub('%obj%', x, '<span style="font-family:\'Times New Roman\', Times, serif;font-style: italic">%obj%</span>')}
```

# Abstract

`r phyloprofile()` contains various analysis methods and visualization of phylogenetic profiles. 

## Preprocess

+ Retrieving phylogenetic profile from the [STRING](http://string-db.org/) database
+ Normalization 
    * Normalized Phylogenetic Profile (NPP)
    * Singular Value Decomposition (SVD)
   
## Supported analysis

+ Co-occurrency methods
    * Jaccard similarity
    * Pearson correlation coefficient
    * mutual information
    * Hamming distance
    * Euclidean distance
    * co-occurrency method with collapsed tree
+ Model-based methods
    * Maximum likelihood
    * Dollo parsimony distance
+ Downstream analysis
    * Expanding known gene-sets
    * Significant gene-set analysis
    
## Visualization

+ Phylogenetic profile plot
+ correlation matrix plot
+ [Circos](http://circos.ca/) plot
+ [D3](https://d3js.org/) network

# Introduction

Phylogenetic profiling is a bioinformatic technique used to show the presence and absence of homologous across a large number of species. This genomic-context method is based on the assumption that functionally related proteins are likely to been evolved in a correlated pattern [@pellegrini1999assigning]. It has been successfully applied to detect novel physical protein-protein interactions, new components of biological complexes, and potential linkages in signaling pathways or metabolic processes [@kensche2008practical].

As increasing numbers of sequenced genomes have become available, various phylogenetic profiling online tools and databases have been developed [@cheng2015protphylo; @sadreyev2015phylogene; @ott2012napp; @szklarczyk2014string; @cromar2016phylopro2]. However, systematically analysis of user-defined phylogenetic profile on local computers is still a slow and tedious process. Here, we provide a R package named `r phyloprofile()` for analyzing and visualizing phylogenetic profiles.

# Construct phylogenetic profiles

In `r phyloprofile()`, a profile with $n$ proteins across $m$ organisms is denoted as a $n \times m$ matrix containing either binning or continuous data. A binning profile uses 0 and 1 to represent the homologous absence and presence, respectively. For example, if the value in the $i$-th row and $j$-th column is 1, it tells the $i$-th protein has corresponding homologous in the $j$-th organism. A continuous profile usually consists of specific sequence similarities, for example BLAST bit scores, which need to be normalized before further analysis. 

## Read profiles from local files

Normally, a profile can be easily stored as a numeric matrix. In `r phyloprofile()` a new class named `r Rclass('PP')` is created to manipulate a profile, and `PP()` is designed to coerce a regular matrix to a `r Rclass('PP')` object. The unique row names of a profile representing protein IDs are particularly important, because the following analysis use them as indices. If row names are missing, `PP()` will automatically creates them, *e.g.* `protein1` and `protein2`, however meaningful names such as [ensembl ID](http://useast.ensembl.org/index.html) or [NCBI gene ID](https://www.ncbi.nlm.nih.gov/gene) are highly recommended. Here, we read in a published profile [@barker2007constrained] called `r Robject('secP')`, which is originally stored in a csv-formed file. In this vignette, we used some "pipe-like" operators, *e.g.* `%>%` and `%<>%` from the `r CRANpkg('magrittr')` package to improve readability of 
codes.

```{r}
library('magrittr')
library('PhyloProfile')

pPath <- system.file('extdata', 'bioinfoProfile.csv', package = 'PhyloProfile')
sceMat <- pPath %>% read.csv(row.names = 1) %>% as.matrix
sceP <- PP(sceMat)

## binning profile
sceP
```

The `r Rclass('PP')` class is directly inherited from the `r Rclass('matrix')` class, so the extractor `[` and `rownames()`/`colnames()` work well for a `r Rclass('PP')` object. Moreover, `PPData()` is used to extract and replace the profile data.


``` {r eval=FALSE}
## profile of top ten proteins
sceP[1:10, ]

## query rownames
rownames(sceP)

## extract and replace profile
testP <- sceP
PPData(testP)
PPData(testP) <- sample(0:1, 5 * 20) %>% matrix(nrow = 5)
```

## Retrieve profiles from the STRING database

The [STRING](http://string-db.org/) database (version 10) provides raw BLAST bit scores of the best hit for more than 2,000 organisms, and it is perfectly suitable for generating raw bit score profiles with the `r Biocpkg('STRINGdb')` package. The `get_STRING_species()` gives a glance of supported organisms in STRING.

```{r}
library('STRINGdb')
sp <- get_STRING_species(version = '10', species_name = NULL)
head(sp)
```

The bit score profile that is a continuous profile can be obtained by `PPSTRING()` according to the taxonomy ID (`r Robject('species_id')` in the above outputs), for example `r Robject('9606')` for *Homo sapiens*. In the following example, we retrieve the profile of 5 human proteins. Set `idx = 'all'` for a complete human profile, and increase `n` to accelerate the download process if multiple cores are supported.

``` {r eval=FALSE}
## A continuous profile
PPSTRING(speID = 9606, idx = 1:5, n = 2)
```

## Normalize raw bit score profiles

Two normalization methods, SVD[@franceschini2015svd; @psomopoulos2013detection] and NPP[@tabach2013identification; @tabach2013human; @sadreyev2015phylogene], are provided for the pre-process of raw bit score profiles.

The algorithm of NPP for a $n \times m$ profile $P$:

1. For each bit score, it is set as `bitReset` if it lower than the `bitCutoff`.

2. For each protein (row), if its number of homologous (with bit score > `bitReset`) across $m$ organisms is lower than `mimConserve`, the protein is removed because of its poor conservation.

3. Normalizing proteins: for the $i$-th row, the bit score $p_{ij}$ is normalized as $\
log_2(p_{ij}/p_{maxi})$, where $p_{maxi}$ is the maximum bit score in the $i$-th row.

4. Normalizing species (columns): for the $j$-th column, the bit score $p_{ij}$ is normalized as $(p_{ij} - \mu_j)/\sigma_j$, which is also known as the [z-score](https://en.wikipedia.org/wiki/Standard_score), where $\mu_j$ and $\sigma_j$ is the mean and the standard deviation of the $j$-th column, respectively.

The algorithm of SVD for a $n \times m$ profile $P$:

1. The first step is the same as the one in NPP.

2. Normalizing proteins: for the $i$-th row, the bit score $p_{ij}$ is normalized as $p_{ij}/p_{maxi}$, where $p_{maxi}$ is the maximum bit score in the $i$-th row.

3. [SVD](https://en.wikipedia.org/wiki/Singular_value_decomposition) of the profile that $\boldsymbol{P = U{\Sigma}V'}$, where $\boldsymbol{U}$ is the unitary matrix. The profile $P'$ then is defined as the top `trimming` (30% for example) columns of $\boldsymbol{U}$. 

4. Similar to the second step of NPP, poor conserved proteins are marked in the original profile $P$, and removed in $P'$.

5. [Euclidean normalization](https://en.wikipedia.org/wiki/Norm_(mathematics)) of species in $P'$: for the $j$-th column, the bit score $p_{ij}$ is normalized as $\sqrt{\sum_{j=1}^{n'}p_{ij}^2}$.


``` {r eval=FALSE}
## simulation of a bit score profile with 2000 proteins and 1000 organisms
ppRawBit <- sample(0:200, 2000 * 1000, replace = TRUE) %>% matrix(nrow = 2000) %>% PP

## SVD
Norm(ppRawBit, method = 'SVD', bitCutoff = 100, bitReset = 0, minConserve = 12, trimming = 0.8)

## NPP
Norm(ppRawBit, method = 'NPP', bitCutoff = 60, bitReset = 1, minConserve = 12)
```

# Linkage analysis

## Prepare linkages

The basic phylogenetic profiling is to analyze the relationships of paired proteins, which is also known as "linkages". `PPIdx()` is used to prepare pre-defined linkages and store linkage indices into a `PPIdx` class object.

``` {r}
## randomly generate 10 linkages
linkM <- sceP %>% rownames %>% sample(2 * 10) %>% matrix(ncol = 2)
sceL <- PPIdx(sceP, linkM)
sceL
```

For a batch of interested proteins, `PPIdx()` also facilities to generate linkages within these proteins or linkages with the rest whole proteome.

``` {r eval=FALSE}
## indices of interested proteins, for example 5 random proteins from `sceP`
p <- c("6321976", "6321639", "6323007", "6320376", "10383799")
pIdx <- p %>% match(rownames(sceP))

## within linkages
PPIdx(sceP, pIdx, pIdx)
## with self linkages, for example "6321976" --> "6321976"
PPIdx(sceP, pIdx, pIdx, self = TRUE)

## linkages with rest whole proteome 
PPIdx(sceP, pIdx, 1:nrow(sceP), bigmat = TRUE)
```

## Co-occurrency methods

Five common-used co-occurrency methods are supported to calculated similarity or distance of linkages.

``` {r}
## mutual information
SimDist(sceL, 'SimMI', n = 2)
```

## Model-based methods



# Case studies



# Misc.

## Convert STRING protein ID

Protein ID (row names) of profiles retrieved from the STRING database are stored in a special format `r Robject('specieID.EnsemblID')`. The following scripts use the `r Biocpkg('biomaRt')` package to convert STRING IDs to other formats such as Entrez ID.

```{r eval=FALSE}
library('PhyloProfile')
library('magrittr')
library('biomaRt')

## STRING profile
hsa5p <- PPSTRING(speID = 9606, idx = 1:5, n = 2)

## step 1: extract Esembl ID
esemblIDs <- hsa5p %>%
  rownames %>%
  strsplit(split = '.', fixed = TRUE) %>%
  sapply('[', 2)

## step 2: check supported formats
ensembl <- useMart('ensembl', dataset = 'hsapiens_gene_ensembl')
listFilters(ensembl)

## step 3: convert to entrez ID
Anno <- getBM(attributes = c('ensembl_peptide_id', 'entrezgene'),
              filters = 'ensembl_peptide_id',
              value = esemblIDs,
              mart = ensembl)
```

## Objects and methods in *PhyloProfile*


# Homepage

Please visit [PhyloProfile homepage](https://github.com/YulongNiu/PhyloProfile) for more information.


# Session Information

Here is the output of `sessionInfo()` on the system on which this document was compiled:

```{r echo=FALSE}
sessionInfo()
```

# References
